cql3/restrictions/statement_restrictions.cc

```diff
@@ -1340,5 +1340,75 @@ const single_column_restrictions::restrictions_map& statement_restrictions::get_
     return single_restrictions->restrictions();
 }
 
+void statement_restrictions::prepare_indexed(const schema& idx_tbl_schema, bool is_local) {
+    if (is_local || !_partition_range_is_simple) {
+        return;
+    }
+    // This assumes that a base-table partition column cannot be the indexed column:
+    _idx_tbl_ck_prefix = std::vector<expr::expression>(1 + _schema->partition_key_size());
+    _idx_tbl_ck_prefix->reserve(idx_tbl_schema.clustering_key_size());
+    for (const auto& e : _partition_range_restrictions) {
+        const auto col = std::get<column_value>(*find(e, oper_t::EQ)->lhs).col;
+        const auto pos = _schema->position(*col) + 1;
+        (*_idx_tbl_ck_prefix)[pos] = replace_column_def(e, &idx_tbl_schema.clustering_column_at(pos));
+    }
+    const column_definition& indexed_column = idx_tbl_schema.column_at(column_kind::partition_key, 0);
+    for (const auto& e : _clustering_prefix_restrictions) {
+        if (find_atom(_clustering_prefix_restrictions[0], expr::is_multi_column)) {
+            // TODO: We could handle single-element tuples, eg. `(c)>=(123)`.
+            break;
+        }
+        const auto any_binop = find_atom(e, [] (auto&&) { return true; });
+        if (!any_binop) {
+            break;
+        }
+        const auto col = std::get<column_value>(*any_binop->lhs).col;
+        if (*col == indexed_column) {
+            continue;
+        }
+        _idx_tbl_ck_prefix->push_back(replace_column_def(e, idx_tbl_schema.get_column_definition(col->name())));
+    }
+    (*_idx_tbl_ck_prefix)[0] = binary_operator(
+            column_value(&idx_tbl_schema.clustering_column_at(0)),
+            oper_t::EQ,
+            // TODO: This should be a unique marker whose value we set at execution time.  There is currently no
+            // handy mechanism for doing that in query_options.
+            ::make_shared<constants::value>(raw_value::make_unset_value()));
```

Comment:
I think this just shows that a this isn't a clustering key prefix, it's a filtering expression. It's not possible to create a slice for indexing, because the token column at the front (if I'm remembering correctly) makes it impossible to slice.

It's fine (and correct) to go this way now as a minimalistic translation, but in the future we may want to split the restrictions into "this goes into partition key range selection", "this goes into clustering key slice creation" and "this is just filtering, be it on the partition key, clustering key, or other column" and this stuff would all go into the filtering section. It could be a single expression (since filtering doesn't need to preserve the clustering key structure).

Reply:
And do mention here that it's the token (if it's the token), this is so opaque.

Reply:
How is this not a CK prefix on the index table?  We just don't know the RHS until we know the token.

CC-ing @eliransin, who I know holds a similar opinion...

Reply:
> And do mention here that it's the token (if it's the token), this is so opaque.

Ah, right, sorry!  I had the same `token_column = ...` line here copied from below, then I decided to be clever and inline it. :facepalm:  I'll bring it back.

Reply:
If the first element in the tuple is unknown, it's not a prefix. And the token is unknown to the user.

e.g. consider

```sql
CREATE TABLE tab (pk int, ck1 int, ck2 int, c int, blah int, PRIMARY KEY (pk, ck1, ck2);

CREATE INDEX ON tab(c);

SELECT * FROM tab WHERE c = ? AND ck1 = ?;
```

This translates to 

```sql
CREATE TABLE idx (c int, _token bigint, pk int, ck1 int, ck2 int, PRIMARY KEY (c, _token, pk, ck1, ck2));

SELECT * FROM idx WHERE c = ? AND ck1 = ?
```

As the user didn't supply _token, this isn't a clustering key prefix.

/cc @psarna (but on vacation)

Reply:
> SELECT * FROM idx WHERE c = ? AND ck1 = ?

This query doesn't use the index, though.
https://github.com/scylladb/scylla/blob/7a5111c580d97935374c55838182a4261b4c8f6a/cql3/statements/select_statement.cc#L1154-L1157

Reply:
Then in what way can we use the clustering key prefix?

I guess "Only EQ restrictions on base partition key can be used in an index view query" refers to local indexes.

Reply:
> I guess "Only EQ restrictions on base partition key can be used in an index view query" refers to local indexes.

No, that whole function is for global indices.

Reply:
> Then in what way can we use the clustering key prefix?

To extend the index-table CK prefix after the token column and all the base-table PK columns.

Reply:
I don't understand. If the user can only specify the column on which we are indexing, what use can we make of the clustering key prefix? No one can supply a value for the first element, the token, so how can we create a clustering slice other than a full slice?

Reply:
Sorry, I didn't explain it properly.  The index is always used when there is an equality restriction on the indexed column (regardless of other restrictions).  If those other restrictions supply enough info to calculate the token, then we restrict the slice.  Otherwise, we fetch the full slice, as you surmised, and later filter the rows fetched.

This new code wrongly presumes that it won't be called when full slice must be fetched.  It also wrongly presumes a partition column is never indexed.  Clearly we need more testing coverage in v3.

Reply:
Ah, of course. If all of the partition key is provided, the token can be computed. It's not necessary for the user to provide it.

Reply:
> This new code wrongly presumes that it won't be called when full slice must be fetched. It also wrongly presumes a partition column is never indexed.

On further examination, neither presumption is wrong.  The second is explained [above](https://github.com/scylladb/scylla/pull/8950#discussion_r662934334); the first is correct because statement_restrictions::get_global_index_clustering_ranges() is only called in this context:
https://github.com/scylladb/scylla/blob/7a5111c580d97935374c55838182a4261b4c8f6a/cql3/statements/select_statement.cc#L1157

If the full PK isn't known, the partition_slice_builder defaults to an open-ended clustering range (ie, full slice).  It's only when the PK is fully specified that this code is called, to construct a set of clustering_ranges to fetch from the index table.

Therefore, this code is correct as it stands.